### ISSUE TITLE
User-define ".DollarNames" does not trigger autocomplete in RStudio

### ISSUE THREAD

**Issue by yuhenghuang** [rstudio/rstudio#7487]
### System details

    RStudio Edition : Desktop
    RStudio Version : 1.3.1056
    OS Version      : Windows 10/Linux (Ubuntu20.04)/ MacOS
    R Version       : 4.0.2

### Steps to reproduce the problem

I was trying the activate the auto-completion triggered by `$` (and press <tab>) for a user-define class. The package I used was `R.oo`, [package page](https://cran.r-project.org/web/packages/R.oo/index.html), which is supposed to be a S3 object.

According the documents, defining the `.DollarNames` for the class correctly would do the job. And it is also noted in the update logs of RStudio that this  `.DollarNames` auto-completion feature has been supported since 1.0.

This is the shortest code that reproduces the problem.

```r
library(R.oo)
library(rlang)

setConstructorS3("List", function(...) {
  extend(
    Object(),
    "List",
    .ls = list(...)
  )
})

setMethodS3(".DollarNames", "List", function(this, pattern="", ...) {
  grep(pattern, names(this[['.ls']]), value = TRUE)
})

setMethodS3("$", "List", function(this, name) {
  key_sym <- ensym(name)
  key <- as_string(key_sym)
  this[['.ls']][[key, exact=TRUE]]
})


ls <- List(abc=1, def=2, hij=3)

ls$abc
```

### Describe the problem in detail

The auto-completions just doesn't work in RStudio. I happened to have PCs at hand with three OSs, while neither of them worked.

However, if I start R from terminal and source the R script, the auto completion triggered by tab worked perfectly.

### Describe the behavior you expected

It will be very helpful if it's possible to enable auto-completion in RStudio for user-define class (specifically `R.oo`) along with proper `.DollarNames` method. I am not sure if this is only some issue related to `R.oo` and if other classes have the same issue.

As the auto-completion itself worked in terminal perfectly, it is most likely that RStudio should support it as well.

- [x] I have read the guide to submitting good bug reports at https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports . 
- [x] I have installed the latest version of RStudio and confirmed that the issue still persists.
- [ ] If I am reporting a RStudio crash, I have included a diagnostics report. https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by ronblum** [comment]
@yuhenghuang Thank you for filing the issue! I can reproduce it in RStudio Desktop 1.4.640 on MacOS 10.15.6 with R 4.0.2. However, it works in the R gui. We'll review this as part of our ongoing development of RStudio.

**Comment by yuhenghuang** [comment]
I just ran into the solution without updating anything that solved the problem purely by chance this morning.
There is actually another necessary condition for the auto-complete to work in RStudio. The mechanism is still unknown and that kind of bugs me. But anyway it definitely works...

Define `names()` method for the class.

In the example I submitted above, adding the following method solved the problem. Hopefully this will be useful for anyone who visited here.

```
setMethodS3("names", "List", function(this) {
  return(names(this[[".ls"]]))
})
```
